### ISSUE TITLE
Missing proxy for Garruk caller of beasts has varying text size; dark black art.

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#588]
![screenshot00421](https://cloud.githubusercontent.com/assets/2443519/11983588/5fbe05b8-a9ac-11e5-9e17-cb32939a873e.png)

**Comment by ShawnieBoy** [comment]
It does look a bit strange. This is due to the separate boxes in the text box resizing to fit - I could try to set them all to the smallest value needed.

As for the image, tokens and planeswalkers don't yet have images yet due to image panels being of non-standard size - even default images.

**Comment by lodici** [comment]
> It does look a bit strange. This is due to the separate boxes in the text box resizing to fit - I could try to set them all to the smallest value used.

On the actual card it is all tiny text. In the above image option 2 likes like a recommendation!

> As for the image, tokens and planeswalkers don't yet have images yet due to image panels being of non-standard size - even default images.

Can you not make an image up like with the other missing images? Or is it the fact that the space is not quite square?

**Comment by ShawnieBoy** [comment]
I can - and it is in the works - it's mainly as the images for those are actually larger than normal crop images, and also require a transparency overlay for the curved edges.

**Comment by ShawnieBoy** [comment]
Will have to adjust the method for writing text to multiple boxes - this issue will no doubt pop up with Leveller cards as well.

Currently each mini-textbox is drawn independently, repeatedly reducing the font-size until it can all fit.

The new method would have to attempt them all, then record the smallest font used, then do it again with that font-size

**Comment by lodici** [comment]
> The new method would have to attempt them all, then record the smallest font used, then do it again with that font-size

That doesn't sound very efficient, but I can't suggest a better solution off the top of my head. There must be a way but I imagine it involves lots of complicated maths.

**Comment by ShawnieBoy** [comment]
Hehe, most likely - That was the first method off the top of my head. Was a real headache getting that to work in the first place, especially with attributed strings.

**Comment by melvinzhang** [comment]
I think the assumption of equal size boxes for each ability is incorrect, if you look at the actual cards, they use different size boxes. Ability with more text uses a larger box. 

In our case, I think we should render the all the ability text into a single "box", the note down where are the separate sections, then place the loyalty counter on the left of each section. This way it will be all the same font size, and each ability get an appropriately sized section. Unfortunately on the frames, the watermark behind the text box assumes equal height so the text will not line up exactly but I think it is more important for the text to be legible then to match the frame exactly.

**Comment by melvinzhang** [comment]
Code wise, drawPlaneswalkerOracleText would use only one box (the entire text area) to render the text, then return an array with the y-bounds of each ability/paragraph. drawLoyaltyPanels would take these y-bounds as input to draw the loyalty panels at the correct location.

**Comment by ShawnieBoy** [comment]
The difficult part there would be determining where the ability breaks are in x,y co-ordinates - and that not all sections will have loyalty values.

Using the generalisation that the majority of planeswalkers have even panels - and have them function similarly to leveller cards, tries to simplify the process down a bit.

As the main issue is with some text appearing larger than others - this could be fixed by reading the length of the longest ability string and rendering that first, and remembering the font size used. As the dimensions are the same, the other abilities can then be rendered at that font size.

**Comment by melvinzhang** [comment]
I think we can assume one ability = one paragraph, i.e. they are separated from each other by line breaks, so far all planeswalker cards have this property. The drawTextToBox knows where the breaks are, it can return them in an array, perhaps something like y-height of first para, y-height of second para. Eg
```
    >
y1  >
    >          
y2  *
    *
y3  $
    $
```

Whether or not there is a loyalty value is already handled by the current code, there is no change to that. The change is that the y-position of the loyalty panel would depend on the y-position of the ability text.

Making them all as small as the smallest font is not ideal, the text rendering doesn't work too well for small sizes. In that case, we would be trading readability for "same font size". They should be as big as they can all fit in the text area.

Examples of problematic images:
![gideon champion of justice](https://cloud.githubusercontent.com/assets/1620634/12078302/4a542952-b247-11e5-9c89-415404bb0b6f.png)
![gideon battle-forged](https://cloud.githubusercontent.com/assets/1620634/12078304/5b233282-b247-11e5-842a-fc233cd2296c.png)

**Comment by ShawnieBoy** [comment]
Yes, I agree with some of the font-sizing problems when it comes to heavy text - this isn't just a planeswalker issue.

Just a comment - those renders look really 'bitty' when it comes to the text, are they taken from in-game?
(Mine as comparison)
![2016-01-03_103344](https://cloud.githubusercontent.com/assets/2363180/12078351/b1ddab30-b205-11e5-8306-55555367f845.png)
![2016-01-03_103557](https://cloud.githubusercontent.com/assets/2363180/12078355/cf5c80aa-b205-11e5-8710-192fb4fe91db.png)

**Comment by melvinzhang** [comment]
Text rendering on Linux has its problems.. anyway here is what they would look like if rendered combined.

![gideon champion of justice](https://cloud.githubusercontent.com/assets/1620634/12078364/636d75e0-b249-11e5-91f6-200901d93d67.png)

![gideon battle-forged](https://cloud.githubusercontent.com/assets/1620634/12078366/6e70b056-b249-11e5-9f55-b1d2586b8812.png)

Hmm, now the problem is those that are very short has almost no vertical height which makes it hard to add the loyalty panel to the left.

**Comment by ShawnieBoy** [comment]
> Hmm, now the problem is those that are very short has almost no vertical height which makes it hard to add the loyalty panel to the left.

I know - I've tried all kinds of things in the past - What's here now was a best-compromise.

In RL games, these cards are the ones that you pick up to read!

**Comment by melvinzhang** [comment]
No matter how short they are, it would still be at least one line + spacing above and below. So as long as line height + spacing is large enough, it will be possible to place the loyalty panel. As shown by the first card, the bottom limit of the text area should be higher so the last line isn't so low.

**Comment by melvinzhang** [comment]
Thought this method will look weird on those with only short lines, possible only use it for planeswalker cards with long text.

**Comment by ShawnieBoy** [comment]
I think it's just Gideon, Nahiri and one of the Nissa's have irregular-sized level panels

**Comment by ShawnieBoy** [comment]
...and Daretti and one of the Tezzeret's
